reps/2023-03-15-train-api.md

```diff
@@ -0,0 +1,143 @@
+
+2. Introduce `Dataset.get_preprocessor_lineage()`, which returns the lineage of preprocessors applied to this Dataset. If multiple preprocessors are applied, it will return a `Chain` of the preprocessors. Non-preprocessor operations on the Dataset are ignored. This function will be used by Train to persist fitted preprocessors with checkpoints.
+
+3. Deprecate the following `DatasetConfig` arguments when streaming is enabled: `max_object_store_memory_fraction`, `global_shuffle`, `fit`, `transform`, `randomize_block_order`, and `per_epoch_preprocessor`, as well as the following `Trainer` arguments: `preprocessor`.
+
+## Example:
+
+### Before
+```python
+
+base = ray.data.read_parquet("s3://bucket/etl_output")
+fact_table = ray.data.read_csv("s3://bucket/my.csv")
+
+# Create the preprocessor.
+prep = StandardScaler(["f1", "f2"])
+
+# Create the per-epoch preprocessor.
+per_epoch_prep = RandomNoisePreprocessor()
+
+# Trainer applies preprocessing internally via config.
+trainer = TorchTrainer(
+    model,
+    datasets={
+        "train_ds": train_ds,
+        "fact_table": fact_table,
+    },
+    scaling_config=ScalingConfig(num_workers=4),
+    preprocessor=prep,
+    dataset_config={
+        "train_ds": {
+            "max_object_store_memory_fraction": 0.2,  # Enable streaming.
+            "randomize_block_order": True,
+            "per_epoch_preprocessor": per_epoch_prep,
+        },
+    },
+)
+
+# Checkpoint includes fitted preprocessor.
+best_checkpoint = trainer.fit().checkpoint
+assert best_checkpoint.get_preprocessor() == prep
+```
+
+### After
+```python
+
+base = ray.data.read_parquet("s3://bucket/etl_output")
+fact_table = ray.data.read_csv("s3://bucket/my.csv")
+
+# Fit the preprocessor.
+prep = StandardScaler(["f1", "f2"])
+prep.fit(base)
+
+# Apply base preprocessing.
+train_ds = base.map_preprocessor(prep)
+train_ds.cache()  # Optional: cache the base data in memory.
+
+# Per-epoch preprocessing.
+train_ds = train_ds \
+    .randomize_block_order() \
+    .map_preprocessor(RandomNoisePreprocessor())
+
+# Trainer doesn't know about preprocessing at all.
+trainer = TorchTrainer(
+    model,
+    datasets={
+        "train_ds": train_ds,
+        "fact_table": fact_table,
+    },
+    scaling_config=ScalingConfig(num_workers=4),
+    dataset_config={
+        "train_ds": {
+            "resource_limits": ExecutionResources(
+                object_store_memory=20e9,  # Customized streaming memory limit.
+            ),
+        },
+    },
+)
+
+# Checkpoint includes fitted preprocessor.
+best_checkpoint = trainer.fit().checkpoint
+assert best_checkpoint.get_preprocessor() == prep
+```
+
+While the "after" code is longer, note that all the data processing code is now cleanly separated from the Trainer, which both a conceptual and practical simplification.
+
```

Comment:
Another win is that stats from the fitted preprocessor are now accessible from within the training function.

For example, if I use a OrdinalEncoder, I can then access the number of unique categories for my feature to know how large to make my model embedding.

This was not possible to do in the old API where fitting would happen under the hood.

Reply:
```python
train_ds, val_ds = ds.read_parquet("s3://data").train_test_split(0.2)
ordinal_encoder = OrdinalEncoder()
per_epoch_preprocessor = BatchMapper(lambda x: add_random_noise(x))

def train_func():
   # I cannot access the preprocessor stats inside my training function.
   num_embeddings = ???

trainer = TorchTrainer(train_func, 
    preprocessor=ordinal_encoder, 
    datasets = {"train": train_ds, "val_ds": val_ds}
    dataset_config=
        {"train": DatasetConfig(per_epoch_preprocessor=per_epoch_preprocessor)
        "val": DatasetConfig(transform=False)}
)
trainer.fit()
```